File: jvm/src/main/scala/com/muchq/scraphics/tracer/Image.scala
package com.muchq.scraphics.tracer

case class Image(width: Int, height: Int):
  val data: Array[Array[Color]] = Array.ofDim[Color](width, height)

  def putPixel(x: Double, y: Int, color: Color): Unit =
    val px: Int = (width / 2 + x).toInt
    val py: Int = (height / 2 - y).toInt - 1
    if 0 <= px && px < width && 0 <= py && py < height then
      data(px)(py) = color

File: domains/graphics/libs/tracy_go/image.go
package tracy

type Image struct {
	Width  int
	Height int
	Data   [][]Color
}

func NewImage(width, height int) *Image {
	data := make([][]Color, width)
	for i := range data {
		data[i] = make([]Color, height)
	}
	return &Image{
		Width:  width,
		Height: height,
		Data:   data,
	}
}

func (img *Image) PutPixel(x, y float64, color Color) {
	px := int(float64(img.Width)/2 + x)
	py := int(float64(img.Height)/2-y) - 1
	if px >= 0 && px < img.Width && py >= 0 && py < img.Height {
		img.Data[px][py] = color
	}
}